### PR TITLE
Recursive Explosion SS + ThrowingSS

### PR DESCRIPTION
**Pull request by MLGTASTICa**
## About The Pull Request
Adds recursive explosions
Adds a subsystem for Throwing and one for explosions
All the code is done , fully functional, fairly balanced.

## Why It's Good For The Game
Cool explosion , throwing no longer sleeps

## Testing
Video with the WIP version (allows warping explosions, but they break a bit since the time adjustment for TiDi is off)
https://youtu.be/y97gC5NUhFo
## Changelog
:cl:
refactor: Added recursive explosions
refactor: Added a throwingSS to better handle item throwing
/:cl:

### CODE REVIEW

**Comment by Pink-Chink** [comment]
A rolling explosion with an unknown source annihilated the ship. Players seem to think a landmine caused it. I'll test this separately from the others.

File: bin/build_and_linter.cmd [hunk 1]
```diff
@@ -0,0 +1,2 @@
+@echo off
+call "%~dp0\..\tools\build\build.bat" --wait-on-error build %*
```
Comment (Pink-Chink) [inline bin/build_and_linter.cmd:2]: If you wanna add stuff like this, put it in its own PR.

File: code/game/atoms.dm [hunk 2]
```diff
@@ -152,8 +152,8 @@
 	if(reagents)
 		QDEL_NULL(reagents)
 
-	spawn()
-		update_openspace()
+	//spawn()
```
Comment (Pink-Chink) [inline code/game/atoms.dm:155]:
```suggestion
```

File: code/game/machinery/doors/windowdoor.dm [hunk 3]
```diff
@@ -65,25 +65,27 @@
 		if(istype(bot))
 			if(density && src.check_access(bot.botcard))
 				open()
-				sleep(50)
+				//sleep(50)
```
Comment (Pink-Chink) [inline code/game/machinery/doors/windowdoor.dm:68]: Turn these sleeps into timers. Windoors can't be bumped like this since they just close immediately.

File: code/game/objects/effects/effect_system.dm [hunk 4]
```diff
@@ -392,17 +407,19 @@ steam.start() -- spawns the effect
...
-				direction = pick(alldirs)
+		var/obj/effect/effect/smoke/smoke = new smoke_type(location)
+		var/direction
+		if(cardinals)
+			direction = pick(cardinal)
+		else
+			direction = pick(alldirs)
+		var/added_time = 1 SECOND
+		for(i=0, i<pick(0,1,1,1,2,2,2,3), i++)
+			addtimer(CALLBACK(src, PROC_REF(move_smoke), smoke, direction), added_time)
```
Comment (Pink-Chink) [inline code/game/objects/effects/effect_system.dm:418]:
```suggestion
		for(var/j=0, j<pick(0,1,1,1,2,2,2,3), j++)
			addtimer(CALLBACK(src, PROC_REF(move_smoke), smoke, direction), added_time)
```
Spawn() allowed this potential infinite loop to work without breaking the server.

File: code/game/objects/effects/effect_system.dm [hunk 5]
```diff
@@ -151,38 +151,50 @@ steam.start() -- spawns the effect
...
+		var/obj/effect/sparks/sparks = new(location)
+		src.total_sparks++
+		var/direction
+		if(src.cardinals)
+			direction = pick(cardinal)
 		else
-			location = get_turf(loca)
+			direction = pick(alldirs)
+		for(i=0, i<pick(1,2,3), i++)
+			addtimer(CALLBACK(src, .proc/do_spark_movement, sparks, direction), rand(1,5) SECONDS)
```
Comment (Pink-Chink) [inline code/game/objects/effects/effect_system.dm:179]:
```suggestion
		for(var/j=0, j<pick(1,2,3), j++)
			addtimer(CALLBACK(src, PROC_REF(do_spark_movement), sparks, direction), rand(1,5) SECONDS)
```
This one already has an exit for the loop, but just in case.

**Comment by ValoTheValo** [comment]
as i said on the discord, there is a large series of bugs with walls

**Comment by ValoTheValo** [comment]
the explosion power on some items is abit ludicrous. a single hand grenade can destroy an entire room and every single item in it, and if you get hit once with an explosion its almost a guarantee whatever your holding will be entirely destroyed

File: code/modules/projectiles/guns/breacher.dm [hunk 6]
```diff
@@ -245,7 +245,7 @@
 		var/mob/living/carbon/human/H = src
 		H.take_overall_damage(35, 10)
 	else
-		ex_act(2)
+		explosion_act(600, null)
```
Comment (Humonitarian) [inline code/modules/projectiles/guns/breacher.dm:248]: Should deal roughly 100-150 damage to mobs instead like it used to (ex_act(2) did 120 damage, ~60 to roaches), otherwise this allows instantly gibbing antags like blitzshells, or very very quickly removing kaisers.

File: code/modules/projectiles/projectile/beams.dm [hunk 7]
```diff
@@ -124,7 +124,7 @@
 
 /obj/item/projectile/beam/pulse/on_hit(atom/target)
 	if(isturf(target))
-		target.ex_act(2)
+		target.explosion_act(500, null)
```
Comment (Humonitarian) [inline code/modules/projectiles/projectile/beams.dm:127]: Same goes for most other cases of ex_act(2). This for example also gibs humans.

**Review by TheShown911 — APPROVED** [review]
Should be ready by now

**Comment by MLGTASTICa** [comment]
Was merged under humons PR
